File: CSharp/Company/Company.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Company
{
    public class Company
    {
        private readonly IList<Worker> _workers;

        public Company()
        {
            _workers = new List<Worker>();
        }

        public void AddWorker(Worker worker)
        {
            _workers.Add(worker);
        }

        public string GenerateJsonReport()
        {
            // Strategy pattern: Vi endrer oppførsel, slik at visitor vil generere en Json-rapport
            var reportVisitor = new ReportVisitor(new JsonReport());
            return GenerateReport(reportVisitor);
        }

        public string GenerateXmlReport()
        {
            // Strategy pattern: Vi endrer oppførsel, slik at visitor vil generere en Xml-rapport
            var reportVisitor = new ReportVisitor(new XmlReport());
            return GenerateReport(reportVisitor);
        }

        public decimal CalculateYearlyCost()
        {
            var visitor = new YearlyCostVisitor();
            foreach (var worker in _workers)
            {
                worker.Accept(visitor);
            }
            return visitor.YearlyCost;
        }

        public decimal CalculateAverageHourlyCost()
        {
            var visitor = new HourlyCostVisitor();
            foreach (var worker in _workers)
            {
                worker.Accept(visitor);
            }
            return visitor.AvarageHourlyCost;
        }

        private string GenerateReport(ReportVisitor reportVisitor)
        {
            var reportBuilder = new StringBuilder(reportVisitor.GenerateReportHeader());

            foreach (var worker in _workers)
            {
                worker.Accept(reportVisitor);
                reportBuilder.AppendLine(reportVisitor.Report);
            }

            reportBuilder.AppendLine(reportVisitor.GenerateReportFooter());
            return reportBuilder.ToString();
        }
    }
}

File: Java/src/main/java/no/bouvet/vasp/Company.java
package no.bouvet.vasp;

import java.util.ArrayList;
import java.util.List;

public class Company {
	
	private List<Worker> workers;
	
	private ReportStrategy reportStrategy;

	public Company() {
		workers = new ArrayList<>();
	}

	public void addWorker(Worker worker) {
		workers.add(worker);
	}
	
	public void setReportStrategy(ReportStrategy reportStrategy) {
		this.reportStrategy = reportStrategy;
	}
	
	public String generateReport() {
		return reportStrategy.generateReport(workers);
	}

	public double calculateYearlyCost() {
		YearlyCostVisitor visitor = new YearlyCostVisitor();
		workers.forEach(worker -> worker.accept(visitor));
		return visitor.getYearlyCost();
	}

	public double calculateAverageHourlyCost() {
		if (workers.size() == 0)
			return 0.0;

		HourlyCostVisitor visitor = new HourlyCostVisitor();
		workers.forEach(worker -> worker.accept(visitor));
		return visitor.getAverageHourlyCost();
	}
}
